src/utils/FileEmitter.ts

```diff
@@ -353,48 +367,82 @@ export class FileEmitter {
 		return this.assignReferenceId(consumedChunk, emittedChunk.id);
 	}
 
-	private finalizeAsset(
+	private finalizeAdditionalAsset(
 		consumedFile: Readonly<ConsumedAsset>,
 		source: string | Uint8Array,
-		referenceId: string,
 		{ bundle, fileNamesBySource, outputOptions }: FileEmitterOutput
 	): void {
-		let fileName = consumedFile.fileName;
+		let { fileName, needsCodeReference, referenceId } = consumedFile;
 
 		// Deduplicate assets if an explicit fileName is not provided
 		if (!fileName) {
 			const sourceHash = getSourceHash(source);
 			fileName = fileNamesBySource.get(sourceHash);
-			const newFileName = generateAssetFileName(
-				consumedFile.name,
+			if (!fileName) {
+				fileName = generateAssetFileName(
+					consumedFile.name,
+					source,
+					sourceHash,
+					outputOptions,
+					bundle
+				);
+				fileNamesBySource.set(sourceHash, fileName);
+			}
+		}
+
+		// We must not modify the original assets to avoid interaction between outputs
+		const assetWithFileName = { ...consumedFile, fileName, source };
+		this.filesByReferenceId.set(referenceId, assetWithFileName);
+
+		const existingAsset = bundle[fileName];
+		if (existingAsset?.type === 'asset') {
+			existingAsset.needsCodeReference &&= needsCodeReference;
+		} else {
+			bundle[fileName] = {
+				fileName,
+				name: consumedFile.name,
+				needsCodeReference,
 				source,
+				type: 'asset'
+			};
+		}
+	}
+
+	private finalizeAssetsWithSameSource(
+		consumedFiles: ReadonlyArray<ConsumedAsset>,
+		sourceHash: string,
+		{ bundle, fileNamesBySource, outputOptions }: FileEmitterOutput
+	): void {
+		let fileName = '';
+		let usedConsumedFile: ConsumedAsset;
+		let needsCodeReference = true;
+		for (const consumedFile of consumedFiles) {
+			needsCodeReference &&= consumedFile.needsCodeReference;
+			const assetFileName = generateAssetFileName(
+				consumedFile.name,
+				consumedFile.source!,
 				sourceHash,
 				outputOptions,
 				bundle
 			);
-			// make sure file name deterministic in parallel emits, always use the shorter and smaller file name
-			if (
-				!fileName ||
-				fileName.length > newFileName.length ||
-				(fileName.length === newFileName.length && fileName > newFileName)
-			) {
-				if (fileName) {
-					delete bundle[fileName];
-				}
-				fileName = newFileName;
-				fileNamesBySource.set(sourceHash, fileName);
+			if (!fileName || assetFileName < fileName) {
```

Comment:
The `fileName` sort strategy is a bit different from #4912. The previous code always uses the shortest `fileName` while this one doesn't; I think it's better to keep the same strategy, otherwise this would be a breaking change.